Support cursor visibility based on input device type

freemangordon opened the pull request:
The logic is: if we receive motion event from a TS device, hide the cursor,
show it otherwise.

Most of the code was borrowed from Cordia h-d

Signed-off-by: Ivaylo Dimitrov <ivo.g.dimitrov.75@gmail.com>

MerlijnWajer commented:
I learned a thing or two from this. :)

freemangordon commented:
This is h-d counterpart of https://github.com/maemo-leste/libmatchbox2/pull/2
Once we agree this is the way to do it, I'll do proper changelog etc

spinal84 commented on src/main.c:
```diff
@@ -351,13 +356,75 @@ key_binding_func_key (MBWindowManager   *wm,
+      /* get XInput DeviceMotion events */
+      devinfo = XListInputDevices (dpy, &ndev);
+
+      if (ndev > sizeof(eclass))
```
How does sizeof(eclass) corellate with the number of available devices?

freemangordon replied:
right, this should be if (ndev > sizeof(eclass) / sizeof(eclass[0])) or
if (ndev > G_N_ELEMENTS(eclass))

BTW the same should be done for if (ndev > sizeof(xdevice_is_ts))

Thanks for pointing that out

spinal84 commented on src/main.c:
```diff
@@ -351,13 +356,75 @@ key_binding_func_key (MBWindowManager   *wm,
+      XDeviceMotionEvent *mev = (XDeviceMotionEvent *)xev;
+      XID devid = mev->deviceid;
+
+      if (devid < sizeof(xdevice_is_ts))
```
same

freemangordon replied:
right, should have caught this :(

spinal84 commented on src/main.c:
```diff
@@ -81,6 +82,10 @@ static int hd_clutter_mutex_enabled = FALSE;
 
+static int xi_motion_ev_type = -1;
+static int xi_presence_ev_type = -1;
+static gboolean xdevice_is_ts[64];
```
I would just use GArray here. Having 256 bytes of static memory occupied for 8 bytes (at most) significant data is just an overkill. GArray in turn allows you to dynamically resize and quickly access items.

freemangordon replied:
ok, makes sense, will rework it to use GArray

spinal84 commented on src/main.c:
```diff
@@ -351,13 +356,75 @@ key_binding_func_key (MBWindowManager   *wm,
+    atom_touchscreen = XInternAtom (dpy, XI_TOUCHSCREEN, True);
+
+  if (atom_touchscreen != None)
+    {
```
Better to be replace with:
```
if (atom_touchscreen == None)
  return;
```
This will make it easier to read the code.

spinal84 replied:
Or just replace:
```
if (atom_touchscreen == None)
    atom_touchscreen = XInternAtom (dpy, XI_TOUCHSCREEN, True);
```
... with:
```
if (atom_touchscreen == None)
  {
    atom_touchscreen = XInternAtom (dpy, XI_TOUCHSCREEN, True);
    if (atom_touchscreen == None)
      return;
  }
```

spinal84 commented on src/main.c:
```diff
@@ -351,13 +356,75 @@ key_binding_func_key (MBWindowManager   *wm,
+            {
+              XDevice *dev = XOpenDevice (dpy, info.id);
+
+              /* bizzaro macro, to SET arg 2 and 3 */
```
what is bizzaro?

freemangordon replied:
This is the original author comment, I'll leave it here. See https://github.com/Cordia/hildon-desktop/commit/df2dd4a9514504285f04f4a6a26e43440b7346f5#diff-803c5170888b8642f2a97e5e9423d399R843

spinal84 replied:
Sure, if you prefer to leave something that you don't understand. Was it a typo?

MerlijnWajer replied:
Bizzaro is slang, it's not a typo. It's like "bizarre"

freemangordon replied:
anyway, I am removing it in the new version of the PR

spinal84 commented on src/main.c:
```diff
@@ -351,13 +356,75 @@ key_binding_func_key (MBWindowManager   *wm,
+
+          if (info.use == IsXExtensionPointer)
+            {
+              XDevice *dev = XOpenDevice (dpy, info.id);
```
Where is the call to XCloseDevice()?

freemangordon replied:
'The XCloseDevice request makes an input device inaccessible to a client through input extension protocol requests'. So, IIUC we should not XCloseDevice() if we want events to come. However, you have a point here - I think before re-enumerating, currently opened devices shall be closed.

spinal84 replied:
According to manual it should also be done before program terminating:
```
Before terminating, and client that has opened input devices through
the input extension should close them via CloseDevice.
```

freemangordon replied:
Right, but I am not sure this is achievable for h-d. Will see what can be done though.

freemangordon committed:
Support cursor visibility based on input device type

The logic is: if we receive motion event from a TS device, hide the cursor,
show it otherwise.

Most of the code was borrowed from Cordia h-d

Signed-off-by: Ivaylo Dimitrov <ivo.g.dimitrov.75@gmail.com>
